pkg/operator/resource/resourceapply/admissions.go

```diff
@@ -2,76 +2,159 @@ package resourceapply
 
 import (
 	"context"
-
-	admissionv1 "k8s.io/api/admissionregistration/v1"
-	"k8s.io/apimachinery/pkg/api/equality"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	admissionclientv1 "k8s.io/client-go/kubernetes/typed/admissionregistration/v1"
-	"k8s.io/klog/v2"
+	"fmt"
 
 	"github.com/openshift/library-go/pkg/operator/events"
 	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	admissionregistrationclientv1 "k8s.io/client-go/kubernetes/typed/admissionregistration/v1"
+	"k8s.io/klog/v2"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-// ApplyValidatingWebhookConfiguration merges objectmeta, update webhooks.
-func ApplyValidatingWebhookConfiguration(client admissionclientv1.ValidatingWebhookConfigurationsGetter, recorder events.Recorder, required *admissionv1.ValidatingWebhookConfiguration) (*admissionv1.ValidatingWebhookConfiguration, bool, error) {
-	existing, err := client.ValidatingWebhookConfigurations().Get(context.TODO(), required.Name, metav1.GetOptions{})
+const genericCABundleInjectorAnnotation = "service.beta.openshift.io/inject-cabundle"
+
+// ApplyMutatingWebhookConfiguration ensures the form of the specified
+// mutatingwebhookconfiguration is present in the API. If it does not exist,
+// it will be created. If it does exist, the metadata of the required
+// mutatingwebhookconfiguration will be merged with the existing mutatingwebhookconfiguration
+// and an update performed if the mutatingwebhookconfiguration spec and metadata differ from
+// the previously required spec and metadata based on generation change.
+func ApplyMutatingWebhookConfiguration(client admissionregistrationclientv1.MutatingWebhookConfigurationsGetter, recorder events.Recorder,
+	requiredOriginal *admissionregistrationv1.MutatingWebhookConfiguration, expectedGeneration int64) (*admissionregistrationv1.MutatingWebhookConfiguration, bool, error) {
+
+	if requiredOriginal == nil {
+		return nil, false, fmt.Errorf("Unexpected nil instead of an object")
+	}
+	required := requiredOriginal.DeepCopy()
+
```

Comment:
This is one thing which I don't think we should do. That mix of hashing and generation is what was causing the necessity of using server-side apply. If you really require to see the full populated spec of the resource, as the server returns it, you need to copy defaulting methods from the server. Otherwise, you have to merge the new spec with the previous one (which we could avoid). In any case, this brings chaos in the approach and make it actually difficult to maintain the state with other operators doing changes on the webhooks with the mentioned annotation.

If you'd like to force redeployment of the resource, the expected generation should be set to 0. This is also replicated on operator startup, as the actual cache for the generation will be empty.

I'll add a test case that will provide an example.

Reply:
Hash and generation serve different purposes.
* Generation (in `existing`) signalizes that something outside of the operator has modified the WebhookConfiguration and the operator should overwrite these changes.
* Hash of `required` detects changes in `required` - e.g. when a new operator with say new `webhook[].timeout` starts, it should update existing WebhookConfiguration. It does not hash `existing` with defaulted values and caBundle and whatnot! The hash should be very very stable, unless the operator really needs a different `required` than it provided previously. We do the same for Deployments / DaemonSets and they don't mess up the generation at all.

Reply:
There are corner cases. Let's say we just created a new webhookConfiguration and didn't use some of the default values which are populated by the server, instead we set the same on the resource ourselves for "better visibility". Then after some time, we remove those - from the server perspective those resources are absolutely equal, so it should not do anything. But it has to apply a new version just because of the hash.

Reply:
IMO that's expected and OK. Hash is not a perfect solution, it may lead to unnecessary updates. But the update will be just one (to get the hash into annotation), not an endless loop of updates.

Reply:
This just seems like an unnecessary complication. Generation covers this use-case too, and we won't have to copy fields with changed certificates from https://github.com/openshift/service-ca-operator to calculate the hash correctly on each reconcile. The code is less error-prone IMO when we rely on a single source of truth, and generation provides a stable one.

Reply:
No, Generation does not cover this use case at all. And these are not a corner case either.

* Say, you have operator v4.7 that deploys ValidatingWebhookConfiguration with 1 webhook. It gets applied and gets `Generation: 2` and the `operator.Status.Generations[..].LastGeneration: 2`.
* In operator v4.8, you add a new webhook, so the array has now 2 items.
  * `operator.Status` has still `Generations[..].LastGeneration: 2`
  *  ValidatingWebhookConfiguration has `Generation: 2`, which matches the expected one.

-> The current code won't update  the object at all! There will be still only one webhook[]. It will introduce hard to debug issues (trust me, been there...). Ad that's why we added hash of spec to Deployment+DaemonSet and we should add them to *WebhookConfiguration too.

Reply:
In this case, this is managed on a higher level. There is surely at least one restart for the operator, so the leader for handling this state will change. The approach is different from the typical one, but it is still the same from a functional perspective. We just store this "hash" approximation in the controller cache, which makes more sense, as the operator's responsibility is to apply his version of the resource at least once, but not more than that. In the case of the stored hash on the resource, it is worse - the new operator may not ever put his version of resource even on startup if someone would sneak in the correct spec hash, but pass a different spec content while the operator was down/upgrading etc..

Reply:
> as the operator's responsibility is to apply his version of the resource at least once,
Why an operator must apply it at least once? `Apply*WebhookConfig` can see that everything is fine (using this hash) and nothing is needed.

Anyway, I can apply the has in the operator, but I'd bet some users of Apply* will get it wrong.